File: Backend/src/utils/normalize.js
function normalizeText(text) {
  if (text === null || text === undefined) return "";

  return String(text)
    .toLowerCase()
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^a-z0-9\s]/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function escapeRegex(text) {
  return String(text).replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function uniqueKeepOrder(items) {
  const seen = new Set();
  const out = [];

  for (const item of items) {
    if (!item) continue;
    if (!seen.has(item)) {
      seen.add(item);
      out.push(item);
    }
  }

  return out;
}

function extractSearchTerms(query) {
  const raw = String(query || "").trim();
  const normalized = normalizeText(raw);

  const terms = [];
  if (raw) terms.push(raw);
  if (normalized) terms.push(normalized);

  for (const token of normalized.split(" ")) {
    if (token.length >= 2) terms.push(token);
  }

  return uniqueKeepOrder(terms);
}

function highlightText(text, terms) {
  if (!text) return "";

  const termList = Array.isArray(terms)
    ? uniqueKeepOrder(terms.map((t) => String(t).trim()).filter(Boolean))
    : extractSearchTerms(terms);

  let result = String(text);

  for (const term of termList.sort((a, b) => b.length - a.length)) {
    const safe = escapeRegex(term);
    const re = new RegExp(`(${safe})`, "gi");
    result = result.replace(re, "<mark>$1</mark>");
  }

  return result;
}

module.exports = {
  normalizeText,
  escapeRegex,
  extractSearchTerms,
  highlightText,
  uniqueKeepOrder
};

File: ml-service/app/utils/normalize.py
import re
import unicodedata


def normalize_text(text):
    if text is None:
        return ""
    text = str(text).lower()
    text = unicodedata.normalize("NFKD", text)
    text = "".join(c for c in text if not unicodedata.combining(c))
    text = re.sub(r"[^a-z0-9\s]", " ", text)
    text = re.sub(r"\s+", " ", text).strip()
    return text


def unique_keep_order(items):
    seen = set()
    out = []
    for item in items:
        if item and item not in seen:
            seen.add(item)
            out.append(item)
    return out


def highlight_text(text, terms):
    if not text:
        return ""

    if isinstance(terms, str):
        terms = [terms]

    result = str(text)
    clean_terms = unique_keep_order([str(t).strip() for t in terms if str(t).strip()])
    clean_terms.sort(key=len, reverse=True)

    for term in clean_terms:
        pattern = re.compile(re.escape(term), re.IGNORECASE)
        result = pattern.sub(lambda m: f"<mark>{m.group(0)}</mark>", result)

    return result